stan/math/rev/fun/quad_form.hpp

```diff
@@ -93,6 +94,65 @@ class quad_form_vari : public vari {
 
   quad_form_vari_alloc<Ta, Ra, Ca, Tb, Rb, Cb>* impl_;
 };
+
+/**
+ * Return the quadratic form \f$ B^T A B \f$.
+ *
+ * Symmetry of the resulting matrix is not guaranteed due to numerical
+ * precision.
+ *
+ * @tparam Mat1 type of the first (square) matrix
+ * @tparam Mat2 type of the second matrix
+ *
+ * @param A square matrix
+ * @param B second matrix
+ * @param symmetric indicates whether the output should be made symmetric
+ * @return The quadratic form
+ * @throws std::invalid_argument if A is not square, or if A cannot be
+ * multiplied by B
+ */
+template <typename Mat1, typename Mat2,
+          require_all_matrix_t<Mat1, Mat2>* = nullptr,
+          require_any_var_matrix_t<Mat1, Mat2>* = nullptr>
+inline auto quad_form_impl(const Mat1& A, const Mat2& B, bool symmetric) {
+  check_square("quad_form", "A", A);
+  check_multiplicable("quad_form", "A", A, "B", B);
+
+  using return_t
+      = promote_var_matrix_t<decltype(value_of(B).transpose().eval()
+                                      * value_of(A) * value_of(B).eval()),
+                             Mat1, Mat2>;
+
+  auto arena_A = to_arena(to_var_value_if<!is_constant<Mat1>::value>(A));
+  auto arena_B = to_arena(to_var_value_if<!is_constant<Mat2>::value>(B));
+
+  check_not_nan("multiply", "A", value_of(arena_A));
+  check_not_nan("multiply", "B", value_of(arena_B));
+
+  auto arena_res = to_arena(value_of(arena_B).transpose() * value_of(arena_A)
+                            * value_of(arena_B));
+
+  if (symmetric) {
+    arena_res = (arena_res + arena_res.transpose()).eval();
+  }
+
+  return_t res = arena_res;
+
+  reverse_pass_callback([arena_A, arena_B, res]() mutable {
+    auto C_adj_B_t = (res.adj() * value_of(arena_B).transpose()).eval();
+
+    if (!is_constant<Mat1>::value)
+      forward_as<promote_var_matrix_t<Mat1, Mat1, Mat2>>(arena_A).adj()
+          += value_of(arena_B) * C_adj_B_t;
+
+    if (!is_constant<Mat2>::value)
+      forward_as<promote_var_matrix_t<Mat2, Mat1, Mat2>>(arena_B).adj()
+          += value_of(arena_A) * C_adj_B_t.transpose()
+             + value_of(arena_A).transpose() * value_of(arena_B) * res.adj();
+  });
+
+  return res;
```

Comment:
[optional/rant] Another thing that's nice about moving out the control flow is that it let's you have the immediete evaluation for `C_adj_B_t` when both are var types and for when just one is a var you can do `C_adj_B_t`  inline for the accumulation.

Reply:
Yeah I see what you're saying.

Are we sure there's no temporary created in an expression like `A * B * C` though? I assume that is broken up into `D = B * C` and `A * D` if `A`, `B`, and `C` are all Eigen MatrixXd types.

Reply:
I'm not sure, the godbolt for that is getting at the level that I can read asm. Though it looks like it makes at least one alloc then can reuse that allocate with a memset under certain conditions. The main thing is that

```cpp
A += B * C * D;
```

Needs to make at least one copy for (`B * C * D`) because of aliasing issues with the left and right hand side. Reading over the asm I think it does one alloc here and reuses the memory for the other multiply. And idt we can use `.noalias()` for eigen matrices of vars since `.noalias()` is like the `restrict` keyword and we access the value and adjoint through the same pointer on both sides (aka aliasing)

Though we can use `.noalias()` for var matrices when we do things like

```cpp
A.adj().noalias() += ...
``

Since the values and adjoints of the var matrix are actually different pointers so there is in fact no aliasing. Does the `restrict` thing make sense?

https://godbolt.org/z/h5df65

Reply:
> I'm not sure, the godbolt for that is getting at the level that I can read asm.

Alright I probably shouldn't have asked that question. Imo we're spending too long on this function.

We could do this efficiently like:
```cpp
auto C_adj_B_t = to_ref_if<!is_constant<Mat1>::value && !is_constant<Mat2>::value>
      (res.adj() * value_of(arena_B).transpose());
```

But the bigger question is `to_var_value_if` and the thing here: https://github.com/stan-dev/math/pull/2164#discussion_r515205941 so let's get those worked out